lib/react_client/react_interop.dart

```diff
@@ -34,8 +36,83 @@ abstract class React {
 
   external static bool isValidElement(dynamic object);
   external static ReactClass get Fragment;
+
+  external static JsRef createRef();
+}
+
+/// Creates a [Ref] object that can be attached to an element via the ref attribute.
+///
+/// __Example__:
+///
+///     class FooComponent extends react.Component2 {
+///       final Ref<BarComponent> barRef = createRef();
+///       final Ref<InputElement> inputRef = createRef();
+///
+///       render() => react.div({}, [
+///         Bar({'ref': barRef}),
+///         react.input({'ref': inputRef}),
+///       ]);
+///     }
+///
+/// Learn more: <https://reactjs.org/docs/refs-and-the-dom.html#creating-refs>.
+Ref<CurrentType> createRef<CurrentType>() {
+  return new Ref<CurrentType>();
+}
+
+/// The return type of [createRef], the constructor wraps [React.createRef] for use in
+/// Dart.
+///
+/// A [Ref] can be attached to an element via the ref attribute. When the element renders,
+/// a reference to that node becomes accessible at the [current] attribute of the [Ref].
+///
+/// See: <https://reactjs.org/docs/refs-and-the-dom.html#creating-refs>.
+class Ref<CurrentType> {
+  JsRef jsRef;
+
+  Ref() {
+    jsRef = React.createRef();
+  }
+
+  /// When the element renders, a reference to that node becomes accessible at the
+  /// [current] attribute of the [Ref].
+  ///
+  /// See: <https://reactjs.org/docs/refs-and-the-dom.html#creating-refs>.
+  CurrentType get current {
+    final jsCurrent = jsRef.current;
+
+    if (jsCurrent is! Element) {
+      final dartCurrent = jsCurrent?.dartComponent;
+
+      if (dartCurrent != null) {
+        return dartCurrent;
```

Comment:
If we do a cast above, we need another cast here, since react.Component isn't necessarily a subtype of `CurrentType`:
```suggestion
        return dartCurrent as CurrentType;
```